search/controllers/api/classic_parser.py

```diff
@@ -0,0 +1,190 @@
+                token_start = i+1
+        elif c == '"':
+            in_quote = not in_quote # flip quotation bit
+        elif c == ' ':
+            if in_quote:
+                continue # keep moving if parsing a quote
+            elif not paren_group and token_start != i:
+                tokens.append(query[token_start:i]) # append the token
+                token_start = i + 1
+            elif token_start == i:
+                token_start = i + 1 # multiple spaces, move the token_start back
+        else:
+            continue
+
+    # handle final-position token
+    if query[token_start:]:
+        tokens.append(query[token_start:])
+
+    # handle unclosed quotation mark
+    if in_quote:
+        raise BadRequest(f'Quotation does not close: {query[token_start:]}')
+
+    return tokens
+
+
+def _cast_tokens(tokens: List[Union[str, Phrase]]) -> List[Union[Operator, Term, Phrase]]:
+    """Cast tokens to class-based representations."""
+    classed_tokens: List[Union[Operator, Term, Phrase]] = []
+    for token in tokens:
+        if isinstance(token, str):
+            if ':' in token:
+                token = _parse_field_query(token)
+                classed_tokens.append(token)
+            else:
+                operator = _parse_operator(token)
+                classed_tokens.append(operator)
+        else:
+            classed_tokens.append(token)
+
+    return classed_tokens
+
+
+def _group_tokens(classed_tokens: List[Union[Operator, Term, Phrase]]) -> Phrase:
+    """Group operators together with Term."""
+    phrases: List[Phrase] = []
+    current_op: Optional[Operator] = None
+    prev_token: Optional[Union[Operator, Term, Phrase]] = None
+    for token in classed_tokens:
+        if isinstance(token, Operator) and isinstance(prev_token, Operator):
+            raise BadRequest(f'Query contains 2 consecutive operators: {prev_token} {token}')
+        if isinstance(token, Operator):
+            current_op = token
+        else:
+            if current_op:
+                token = (current_op, token)
+            phrases.append(token)
+        prev_token = token
+
+    # Return single-token query, otherwise cast to a tuple.
+    if len(phrases) == 1:
+        return phrases[0]
+    else:
+        # When mypy adds full support for recursive types, this should be fine.
+        return tuple(phrases)  # type: ignore
+
+def _parse_operator(characters: str) -> Operator:
+    try:
+        return Operator(characters.strip())
+    except ValueError as e:
+        raise BadRequest(f'Cannot parse fragment: {characters}') from e
+
+
+def _parse_field_query(field_part: str) -> Term:
+    field_name, value = field_part.split(':', 1)
+
+    # cast field to Field enum
+    try:
+        field = Field(field_name)
+    except ValueError as e:
+        raise BadRequest(f'Invalid field: {field_name}') from e
+
+    # process quotes, if present
+    if value.startswith('"') and value.endswith('"'):
```

Comment:
what about ``value = ' "a quoted phrase" '``?

Reply:
added a `value = value.strip()` before this block:

```

    # Process leading and trailing whitespace and quotes, if present.
    value = value.strip()
    if value.startswith('"') and value.endswith('"'):
        value = value[1:-1]
```